File: flutter/lib/services/api_client.dart
import 'dart:async';

import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../config/app_config.dart';
import '../models/auth_failure.dart';
import 'token_storage.dart';
// authProvider imported for the lazy `ref.read` inside onError — the
// interceptor closure only reads it later, inside the async callback, well
// after every provider has finished its initial construction. This is not a
// circular dependency even though authProvider -> authApiProvider ->
// dioProvider, because dioProvider never eagerly watches/reads authProvider
// during its own build.
import '../providers/auth_provider.dart';

final dioProvider = Provider<Dio>((ref) {
  final config = ref.read(appConfigProvider);
  final tokenStorage = ref.read(tokenStorageProvider);
  final dio = Dio(BaseOptions(baseUrl: config.apiBaseUrl));

  Completer<String?>? refreshCompleter;

  dio.interceptors.add(
    InterceptorsWrapper(
      onRequest: (options, handler) async {
        final isAuthRoute = options.path.contains('/api/auth/login') ||
            options.path.contains('/api/auth/refresh');
        if (!isAuthRoute) {
          final accessToken = await tokenStorage.readAccessToken();
          if (accessToken != null) {
            options.headers['Authorization'] = 'Bearer $accessToken';
          }
        }
        handler.next(options);
      },
      onError: (error, handler) async {
        final isRefreshCall = error.requestOptions.path.contains('/api/auth/refresh');
        final alreadyRetried = error.requestOptions.extra['isRetry'] == true;

        if (error.response?.statusCode != 401 || isRefreshCall || alreadyRetried) {
          return handler.next(error);
        }

        final isFirstToRefresh = refreshCompleter == null;
        refreshCompleter ??= Completer<String?>();

        try {
          String? newAccessToken;
          if (isFirstToRefresh) {
            final refreshToken = await tokenStorage.readRefreshToken();
            if (refreshToken == null) throw const AuthFailure.invalidCredentials();

            final response = await dio.post(
              '/api/auth/refresh',
              data: {'refreshToken': refreshToken},
            );
            newAccessToken = response.data['accessToken'] as String;
            await tokenStorage.writeAccessToken(newAccessToken);
            refreshCompleter!.complete(newAccessToken);
          } else {
            newAccessToken = await refreshCompleter!.future;
          }

          final retryOptions = error.requestOptions
            ..headers['Authorization'] = 'Bearer $newAccessToken'
            ..extra['isRetry'] = true;
          final retryResponse = await dio.fetch(retryOptions);
          return handler.resolve(retryResponse);
        } catch (e) {
          if (isFirstToRefresh) refreshCompleter!.completeError(e);
          await ref.read(authProvider.notifier).forceLogout();
          return handler.next(error);
        } finally {
          if (isFirstToRefresh) refreshCompleter = null;
        }
      },
    ),
  );

  return dio;
});

File: app/app/services/apiClient.ts
import { tokenStorage } from './tokenStorage'

interface FetchErrorLike {
  response?: { status?: number }
}

let refreshPromise: Promise<string> | null = null

function baseUrl(): string {
  return useRuntimeConfig().public.apiBaseUrl
}

function isAuthRoute(path: string): boolean {
  return path.includes('/api/auth/login') || path.includes('/api/auth/refresh')
}

async function refreshAccessToken(): Promise<string> {
  const refreshToken = await tokenStorage.readRefreshToken()
  if (!refreshToken) throw new Error('no-refresh-token')

  const { accessToken } = await $fetch<{ accessToken: string }>('/api/auth/refresh', {
    baseURL: baseUrl(),
    method: 'POST',
    body: { refreshToken },
  })
  await tokenStorage.writeAccessToken(accessToken)
  return accessToken
}

// Bearer-token attachment + a single deduplicated refresh-and-retry on 401,
// mirroring the Flutter client's Dio interceptor (flutter/lib/services/api_client.dart).
// Import of the auth store is deferred (dynamic import) so this module never
// eagerly pulls in stores/auth.ts, which itself imports authApi — avoiding a
// circular-import cycle at module-load time.
export async function apiRequest<T>(
  path: string,
  options: { method?: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH'; body?: unknown } = {},
  isRetry = false,
): Promise<T> {
  const headers: Record<string, string> = {}
  if (!isAuthRoute(path)) {
    const accessToken = await tokenStorage.readAccessToken()
    if (accessToken) headers.Authorization = `Bearer ${accessToken}`
  }

  try {
    return await $fetch<T>(path, {
      baseURL: baseUrl(),
      method: options.method ?? 'GET',
      body: options.body,
      headers,
    })
  } catch (error) {
    const status = (error as FetchErrorLike).response?.status
    if (status !== 401 || isAuthRoute(path) || isRetry) throw error

    try {
      refreshPromise ??= refreshAccessToken()
      await refreshPromise
    } catch (refreshError) {
      const { useAuthStore } = await import('../stores/auth')
      await useAuthStore().forceLogout()
      throw refreshError
    } finally {
      refreshPromise = null
    }

    return apiRequest<T>(path, options, true)
  }
}
